### ISSUE TITLE
Unable to build managed assemblies on FreeBSD 10.2

### ISSUE THREAD

**Issue by jasonwilliams200OK** [dotnet/coreclr#1633]
While building CoreFX managed assemblies on FreeBSD, I am encountering the following error on "Restoring all packages.." step:

```csh
/root/corefx/packages/Microsoft.Build.Mono.Debug.14.1.0.0-prerelease/lib/Microsoft.CSharp.CurrentVersion.target
s(383,9): error MSB6006: "csc2.exe" exited with code 1.
```

for

```
Microsoft.Win32.Primitives
Microsoft.Win32.Registry
System.Collections.Concurrent
System.Collections.Immutable
System.Collections.NonGeneric
System.Collections.Specialized
System.Private.Uri
System.Diagnostics.Debug
System.ComponentModel.Annotations
System.ComponentModel.EventBasedAsync
System.ComponentModel.Primitives
System.ComponentModel.TypeConverter
System.ComponentModel
System.Composition.AttributedModel

etc. (total 149 failed)
```

To replicate this issue, I have prepared a guide on how to setup FreeBSD 10.2 on Windows 10 HyperV to compile CoreFX: https://gist.github.com/jasonwilliams200OK/6efa7907e66275df2d24 (@sokket, @naamunds, this guide might interest you guys)

It seems like the issue is originating from buildtools (or MSbuild) when running [`<Csc>` task](https://github.com/Microsoft/msbuild/blob/154de6b319ef4a13550f15534b84450914b21623/src/XMakeTasks/Microsoft.CSharp.CurrentVersion.targets#L246-L305). I am not sure how to debug it, as I tried getting variable values on stdout using `<Exec Command="echo '>> $(CscEnvironment)'"/>`, `<Message Text=">> $(CscEnvironment)" />` etc. but failed to output anything..

/cc: @weshaggard, @akoeplinger, @ericstj, @mmitche

**Comment by akoeplinger** [comment]
Can you post the msbuild.log file?

**Comment by jasonwilliams200OK** [comment]
@akoeplinger, sure, here we go: https://gist.github.com/jasonwilliams200OK/5177df88cce4c6890885.

**Comment by akoeplinger** [comment]
@jasonwilliams200OK that seems to be a private gist so I don't have access (same with the FreeBSD guide above).

**Comment by jasonwilliams200OK** [comment]
Oh i updated the link :)

Also,

```
mono -V

Mono JIT compiler version 4.0.1 (tarball Thu Sep 17 03:06:40 UTC 2015)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notification:  kqueue
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        LLVM:          supported, not enabled.
        GC:            sgen
```

**Comment by akoeplinger** [comment]
@jasonwilliams200OK you're running into a Mono bug (`Failed to read past end of stream`) that has been fixed in a subsequent bugfix release. You need a newer Mono.

**Comment by jasonwilliams200OK** [comment]
Oh yes! This is ditto bug as https://github.com/dotnet/roslyn/issues/2844. So does this mean in past four months, FreeBSD didn't get any new Mono release?  :open_mouth: 
cc @ajensenwaud, @smortex

**Comment by jasonwilliams200OK** [comment]
Turned out mono 4.0.3 is available for 2015-Q4 branch. Meanwhile, we can use "latest" repository to get mono 4.0.3 and clang37. Updated the gist guide accordingly: https://gist.github.com/jasonwilliams200OK/6efa7907e66275df2d24#install-software/. 
Now CoreFX is building on FreeBSD.  :tada: 

/cc @josteink, @janhenke, @ghuntley

**Comment by akoeplinger** [comment]
Sweet :) I guess the next step is running the corefx tests?

**Comment by josteink** [comment]
Awesome. I'm curious about those tests too :)

**Comment by jasonwilliams200OK** [comment]
Hacking 101.. Had to make some tweaks to run the tests:

* `nano ./run-test.sh` and replace `$OS` with `Linux` in URL under `create_test_overlay()` because https://www.myget.org/F/dotnet-buildtools/api/v2/package/Microsoft.DotNet.CoreFx.FreeBSD.TemporaryTestHost/0.0.2-prerelease does not exist.

* Downloaded [CoreCLR artifacts](http://dotnet-ci.cloudapp.net/job/dotnet_coreclr_freebsd_debug/) zip: `cd /root; coreclr -o bins.zip "http://dotnet-ci.cloudapp.net/job/dotnet_coreclr_freebsd_debug/lastSuccessfulBuild/artifact/*zip*/archive.zip"` (don't forget the quotes around URL) and then `unzip bins.zip; chmod -R 0777 archive/; cd corefx`.

* Built mscorlib.dll for FreeBSD on Windows (`c:\repos\coreclr\build.cmd freebsdmscorlib Debug`) and scp'd into VM: `scp \repos\coreclr\bin\Product\FreeBSD.x64.Debug\mscorlib.dll root@192.168.0.28:/root/corefx/packages/Microsoft.DotNet.CoreCLR/1.0.0-prerelease/lib/aspnetcore50/`.

* Ran the test:

  ```
  ./run-test.sh \
  --coreclr-bins      ../archive/bin/Product/FreeBSD.x64.Debug \
  --mscorlib-bins      ./packages/Microsoft.DotNet.CoreCLR/1.0.0-prerelease/lib/aspnetcore50 \
  --corefx-tests       ./bin/tests/FreeBSD.AnyCPU.Debug \
  --corefx-native-bins ./bin/FreeBSD.x64.Debug/Native
  ```

...

...

aaaaaaaaand

...

...

tests are failing with lots of these:

```
********

./run-test.sh: line 172:  3844 Abort trap              (core dumped) ./corerun xunit.console.netcore.exe $testDllName -x
ml testResults.xml -notrait category=failing -notrait category=OuterLoop -notrait category=$xunitOSCategory

Running tests in bin/tests/FreeBSD.AnyCPU.Debug//System.Collections.NonGeneric.Tests/dnxcore50
./corerun xunit.console.netcore.exe System.Collections.NonGeneric.Tests.dll -xml testResults.xml -notrait category=faili
ng -notrait category=OuterLoop -notrait category=nonfreebsdtests


Assert failure(PID 3851 [0x00000f0b], Thread: 29385728 [0x1c06400]): !"Recursion in CLRException::GetThrowable"
    File: /mnt/resource/j/workspace/dotnet_coreclr_freebsd_debug/src/vm/clrex.cpp Line: 150
    Image: /root/corefx/bin/tests/FreeBSD.AnyCPU.Debug/System.Collections.NonGeneric.Tests/dnxcore50/corerun

**** MessageBox invoked, title 'corerun - Assert Failure (PID 3851, Thread 29385728/0x1c06400)' ****
  !"Recursion in CLRException::GetThrowable"

/mnt/resource/j/workspace/dotnet_coreclr_freebsd_debug/src/vm/clrex.cpp, Line: 150

Abort - Kill program
Retry - Debug
Ignore - Keep running


Image:
/root/corefx/bin/tests/FreeBSD.AnyCPU.Debug/System.Collections.NonGeneric.Tests/dnxcore50/corerun

********
```

**Comment by josteink** [comment]
Nice. Obviously there's room for improvement here, but failing tests are much better than tests failing to run :)

**Comment by akoeplinger** [comment]
Are there actually some tests passing or do all fail with these asserts?

**Comment by jasonwilliams200OK** [comment]
All of them are failing. Could it be because https://www.myget.org/F/dotnet-buildtools/api/v2/package/Microsoft.DotNet.CoreFx.FreeBSD.TemporaryTestHost/0.0.2-prerelease is unavailable?

**Comment by akoeplinger** [comment]
Yeah, I suppose that's connected somehow, but I don't know how the TemporaryTestHost is built.

You could try just running `./corerun xunit.console.netcore.exe` without any parameters to see if it outputs the help text (to verify xunit even boots up).

Btw. I just saw we're in `dotnet/coreclr`. I suggest you open a new issue in corefx since the original topic of this issue is fixed :)

**Comment by jasonwilliams200OK** [comment]
```
root@:~/corefx/bin/tests/FreeBSD.AnyCPU.Debug/Common.Tests/dnxcore50 # ./corerun xunit.console.netcore.exe

Assert failure(PID 4180 [0x00001054], Thread: 29385728 [0x1c06400]): !"Recursion in CLRException::GetThrowable"
    File: /mnt/resource/j/workspace/dotnet_coreclr_freebsd_debug/src/vm/clrex.cpp Line: 150
    Image: /root/corefx/bin/tests/FreeBSD.AnyCPU.Debug/Common.Tests/dnxcore50/corerun

**** MessageBox invoked, title 'corerun - Assert Failure (PID 4180, Thread 29385728/0x1c06400)' ****
  !"Recursion in CLRException::GetThrowable"

/mnt/resource/j/workspace/dotnet_coreclr_freebsd_debug/src/vm/clrex.cpp, Line: 150

Abort - Kill program
Retry - Debug
Ignore - Keep running


Image:
/root/corefx/bin/tests/FreeBSD.AnyCPU.Debug/Common.Tests/dnxcore50/corerun

********

Abort (core dumped)
```

Yup I realized it late that I opened it in CoreCLR. Anecdotally, this particular exception (`CLRException::GetThrowable`) appears as if its coming from CoreCLR. :D

**Comment by jasonwilliams200OK** [comment]
@akoeplinger, I think the error is due to the fact `System.Globalization.Native` assembly in CoreFX failed to build. See log produced by FreeBSD `truss` utility (which is similar to Linux `strace`): https://gist.github.com/jasonwilliams200OK/2ed55642a0659cb74086 - first line describes how to build this log.

I will try to build it from scratch once https://github.com/dotnet/corefx/pull/3471 gets merged and then start logging issues on CoreFX repo.

**Comment by ellismg** [comment]
@jasonwilliams200OK Do you have ICU installed?  System.Globalization.Native comes from the CoreCLR repo, FWIW.

**Comment by jasonwilliams200OK** [comment]
@ellismg, thanks for catching it. I actually installed it (using `pkg install icu`) but forgot to update the guide. I have updated the guide https://gist.github.com/jasonwilliams200OK/6efa7907e66275df2d24 and mentioned in the change log for Revision 7 (comment under the guide). Revision 8 was also a separate missing bit. :)

---

Regarding CoreCLR, I am in the middle of creating a similar guide: https://gist.github.com/jasonwilliams200OK/06db121ec534821fe541, but unable to finish it because I haven't been able to successfully build CoreCLR on FreeBSD locally yet. Could you (or @mmitche or @steveharter) please assist what dependency am I missing? I tried running ditto command `./build.sh skipmscorlib verbose debug` as [dotnet-ci is using](https://github.com/dotnet/dotnet-ci/blob/7b6ce7a/jobs/dotnet_coreclr_freebsd_debug/config.xml#L71), the CMake configuraiton is exactly the same, but it fails the build at 1% with the following error:

```
In file included from /root/coreclr/src/pal/src/exception/seh.cpp:23:
/root/coreclr/src/pal/src/include/pal/thread.hpp:511:13: error: use of undeclared identifier 'errno'
            errno = dwLastError;
            ^
/root/coreclr/src/pal/src/include/pal/thread.hpp:520:20: error: use of undeclared identifier 'errno'
            return errno;
                   ^
```

Here is the full build log: https://gist.github.com/jasonwilliams200OK/18d9de1f5c78ed513e47. Note that those warning coming from system libs (which don't show up on Linux) will be gone once https://github.com/dotnet/coreclr/pull/1666 is merged.

**Comment by mmitche** [comment]
That is kind of bizarre.  Same version of clang too.  Have you looked at the source?  Is it obvious that errno is defined someplace?

**Comment by jasonwilliams200OK** [comment]
I think the issue is with: https://github.com/dotnet/coreclr/blob/a723811/src/pal/inc/pal.h#L6127-L6137. Seems like `PAL_STDCPP_COMPAT` check is failing and it tries to redefine `errno`.
But why it is not happening on dotnet-ci :question:  :confused:

**Comment by jasonwilliams200OK** [comment]
Update (on CoreFX again):

Reverted the latest package branch workaround (https://github.com/dotnet/coreclr/issues/1633#issuecomment-143603601) in Revision 10, since 2015-Q4 default branch for FreeBSD packages is available. https://gist.github.com/jasonwilliams200OK/6efa7907e66275df2d24. Which means running `pkg install mono` on fresh FreeBSD 10.2 installation will yield `mono -V` == v4.0.3.20 without any modifications! This version has all the fixes we require for CoreCLR as well as Roslyn. :)

**Comment by jasonwilliams200OK** [comment]
Pingback:

The coreclr issue turned out to be https://github.com/dotnet/coreclr/issues/1091, which was later fixed by https://github.com/dotnet/coreclr/pull/1739. 
Now CoreCLR builds on FreeBSD with these steps: https://gist.github.com/jasonwilliams200OK/06db121ec534821fe541.

Note that if you are on Mac and have Parallels installed, then installing FreeBSD VM from ISO is really a piece of cake! :)
